### PR TITLE
ddl: check the limitation when creating multi-valued index 

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/40086

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: parser/parser.y [hunk 1]
```diff
@@ -7325,6 +7328,15 @@ SimpleExpr:
 		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr(ast.JSONUnquote), Args: []ast.ExprNode{extract}}
 	}
 
+ArrayKwdOpt:
+	{
+		$$ = false
+	}
+|	"ARRAY"
```
Comment (qw4990) [inline parser/parser.y:7335]: Better to add some test cases under the parser package for this new syntax.

File: types/typed_array.go [hunk 2]
```diff
@@ -0,0 +1,3 @@
+package types
```
Comment (qw4990) [inline types/typed_array.go:1]: This new file has no license.

File: expression/builtin_cast.go [hunk 3]
```diff
@@ -407,6 +407,71 @@ func (c *castAsDurationFunctionClass) getFunction(ctx sessionctx.Context, args [
...
+	baseBuiltinFunc
+}
+
+func (b *castJSONAsArrayFunctionSig) Clone() builtinFunc {
+	newSig := &castJSONAsArrayFunctionSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+func (b *castJSONAsArrayFunctionSig) evalJSON(row chunk.Row) (res types.TypedArray, isNull bool, err error) {
```
Comment (qw4990) [inline expression/builtin_cast.go:420]: Should we use `evalArray` instead of `evalJSON` as the function name here?

File: expression/builtin.go [hunk 4]
```diff
@@ -479,6 +483,8 @@ type builtinFunc interface {
 	evalDuration(row chunk.Row) (val types.Duration, isNull bool, err error)
 	// evalJSON evaluates JSON representation of builtinFunc by given row.
 	evalJSON(row chunk.Row) (val types.BinaryJSON, isNull bool, err error)
+	// evalArray evaluates JSON representation of builtinFunc by given row.
+	evalArray(row chunk.Row) (val types.TypedArray, isNull bool, err error)
```
Comment (qw4990) [inline expression/builtin.go:487]: I suggest we reuse the existing `evalJSON` instead of introducing a new `evalArray`.

**Review by tangenta — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

File: ddl/index.go [hunk 5]
```diff
@@ -81,7 +81,12 @@ func buildIndexColumns(ctx sessionctx.Context, columns []*model.ColumnInfo, inde
 		if err := checkIndexColumn(ctx, col, ip.Length); err != nil {
 			return nil, false, err
 		}
-		mvIndex = mvIndex || col.FieldType.IsArray()
+		if col.FieldType.IsArray() {
+			if mvIndex {
+				return nil, false, dbterror.ErrNotSupportedYet.GenWithStack("'more than one multi-valued key part per index'")
+			}
+			mvIndex = true
```
Comment (hawkingrei) [inline ddl/index.go:88]: ```mvIndex``` will make people mistakenly think that it means ```move index```.

Comment (xiongjiwei) [reply ddl/index.go:88]: I think it is ok, only in the function scope

**Pull request merged**